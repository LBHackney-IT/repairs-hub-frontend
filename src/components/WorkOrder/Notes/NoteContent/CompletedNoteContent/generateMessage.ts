import { Note } from '../../types'

type PaymentType = 'Bonus' | 'Overtime' | 'CloseToBase' | null

interface WorkOrderOperative {
  name: string
  jobPercentage: number
}

const generateMessage = (
  userComments: string,
  completionDate: Date | null,
  operativesWithPercentages: WorkOrderOperative[],
  paymentType: PaymentType
): string => {
  let note = 'Closed - Completed'

  if (completionDate) {
    note += `: ${completionDate.toLocaleDateString('en-GB', {
      day: '2-digit',
      month: 'long',
      year: 'numeric',
      hour: '2-digit',
      minute: '2-digit',
    })}`
  }

  if (userComments?.trim()) {
    note += ` - ${userComments.trim()}`
  }

  if (operativesWithPercentages.length > 0) {
    const showPercentage = paymentType !== 'Overtime'
    const operativesWithPercentagesValue = operativesAndPercentagesForNotes(
      operativesWithPercentages,
      showPercentage
    )

    note += ` - Assigned operatives ${operativesWithPercentagesValue}`
  }

  if (paymentType) {
    note += ` - ${workOrderNoteFragmentForPaymentType(paymentType)}`
  }

  return note
}

const workOrderNoteFragmentForPaymentType = (
  paymentType: PaymentType
): string => {
  switch (paymentType) {
    case 'Bonus':
      return 'Bonus calculation'
    case 'Overtime':
      return 'Overtime work order (SMVs not included in Bonus)'
    case 'CloseToBase':
      return 'Close to base (Operative payment made)'
    default:
      return ''
  }
}

const operativesAndPercentagesForNotes = (
  operativesWithPercentages: WorkOrderOperative[],
  showPercentages: boolean
): string => {
  return operativesWithPercentages
    .map((x) => (showPercentages ? `${x.name}: ${x.jobPercentage}%` : x.name))
    .join(', ')
}

export default generateMessage
