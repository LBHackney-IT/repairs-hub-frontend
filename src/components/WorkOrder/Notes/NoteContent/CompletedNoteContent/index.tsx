import { Note, WorkOrderRequest } from '../../types'
import generateMessage from './generateMessage'

interface Props {
  note: Note
  workOrder: WorkOrderRequest
}

const CompletedNoteContent = ({ note, workOrder }: Props) => {
  const completionDate = new Date(workOrder.closedDated)
  
  const statusMessage = generateMessage(
    note.note,
    completionDate,
    workOrder.operatives,
    workOrder.paymentType
  )

  return (
    <>
      <span>
        {statusMessage}
        <br />
      </span>

      {Object.prototype.hasOwnProperty.call(workOrder, 'followOnRequest') &&
        workOrder.followOnRequest !== null && (
          <>
            <div style={{ background: '#eee', padding: '15px' }}>
              <span>
                <strong style={{ fontWeight: 700, color: '#333' }}>
                  Details of further work required
                </strong>{' '}
                <br />
                <br />
              </span>

              <span>
                <strong>Type of work required</strong>
              </span>

              <ul
                className="lbh-list lbh-list--bullet lbh-body-s"
                style={{ color: '#333' }}
              >
                {workOrder.followOnRequest.isSameTrade && (
                  <li style={{ marginTop: '5px' }}>Same trade</li>
                )}
                {workOrder.followOnRequest.isDifferentTrades && (
                  <li style={{ marginTop: '5px' }}>
                    Different trade(s):{' '}
                    {workOrder.followOnRequest.requiredFollowOnTrades.join(
                      ', '
                    )}
                  </li>
                )}
                {workOrder.followOnRequest.isMultipleOperatives && (
                  <li style={{ marginTop: '5px' }}>Multiple operatives</li>
                )}
              </ul>

              {workOrder.followOnRequest.followOnTypeDescription !== null &&
                workOrder.followOnRequest.followOnTypeDescription.trim()
                  .length > 0 && (
                  <>
                    <br />
                    <span style={{ color: '#333' }}>
                      {workOrder.followOnRequest.followOnTypeDescription}
                      <br />
                    </span>
                  </>
                )}

              {(workOrder.followOnRequest.stockItemsRequired ||
                workOrder.followOnRequest.nonStockItemsRequired ||
                (workOrder.followOnRequest.materialNotes !== null &&
                  workOrder.followOnRequest.materialNotes.trim().length >
                    0)) && (
                <>
                  <br />
                  <span>
                    <strong>Materials required</strong>
                  </span>

                  {(workOrder.followOnRequest.stockItemsRequired ||
                    workOrder.followOnRequest.nonStockItemsRequired) && (
                    <ul
                      className="lbh-list lbh-list--bullet lbh-body-s"
                      style={{ color: '#333' }}
                    >
                      {workOrder.followOnRequest.stockItemsRequired && (
                        <li style={{ marginTop: '5px' }}>
                          Stock items required
                        </li>
                      )}

                      {workOrder.followOnRequest.nonStockItemsRequired && (
                        <li style={{ marginTop: '5px' }}>
                          Non stock items required
                        </li>
                      )}
                    </ul>
                  )}

                  {workOrder.followOnRequest.materialNotes !== null &&
                    workOrder.followOnRequest.materialNotes.trim().length >
                      0 && (
                      <>
                        <br />
                        <span style={{ color: '#333' }}>
                          {workOrder.followOnRequest.materialNotes}
                          <br />
                        </span>
                      </>
                    )}
                </>
              )}

              {workOrder.followOnRequest.additionalNotes !== null &&
                workOrder.followOnRequest.additionalNotes.trim().length > 0 && (
                  <>
                    <br />

                    <span>
                      <strong>Additional notes</strong>
                      <br />
                    </span>

                    <span style={{ color: '#333' }}>
                      {workOrder.followOnRequest.additionalNotes}
                    </span>
                    <br />
                  </>
                )}
            </div>
          </>
        )}

      {/* To do - add back later */}
      {/* <span>
        <a href="#photos-tab">2 photos uploaded</a> <br />
      </span> */}
    </>
  )
}

export default CompletedNoteContent
