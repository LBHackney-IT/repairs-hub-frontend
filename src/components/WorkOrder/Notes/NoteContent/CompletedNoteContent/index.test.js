import { render } from '@testing-library/react'
import CompletedNoteContent from '.'

describe('CompletedNoteContent component', () => {
  it('should match snapshot when user comment is empty', () => {
    const workOrder = {
      closedDated: '2024-09-26T12:34:56',
      paymentType: 'Bonus',
      operatives: [{ name: 'Operative Name', jobPercentage: 100 }],
    }
    const note = { note: '' }

    const { asFragment } = render(
      <CompletedNoteContent
        note={note}
        workOrder={workOrder}
      />
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('should match snapshot when user comment is not empty', () => {
    const workOrder = {
      closedDated: '2024-09-26T12:34:56',
      paymentType: 'Bonus',
      operatives: [{ name: 'Operative Name', jobPercentage: 100 }],
    }
    const note = { note: 'User comment' }

    const { asFragment } = render(
      <CompletedNoteContent
        note={note}
        workOrder={workOrder}
      />
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('should match snapshot when no operative', () => {
    const workOrder = {
      closedDated: '2024-09-26T12:34:56',
      paymentType: 'Bonus',
      operatives: [],
    }
    const note = { note: 'User comment' }

    const { asFragment } = render(
      <CompletedNoteContent
        note={note}
        workOrder={workOrder}
      />
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('should match snapshot when one operative', () => {
    const workOrder = {
      closedDated: '2024-09-26T12:34:56',
      paymentType: 'Bonus',
      operatives: [{ name: 'Operative Name', jobPercentage: 50 }],
    }
    const note = { note: 'User comment' }

    const { asFragment } = render(
      <CompletedNoteContent
        note={note}
        workOrder={workOrder}
      />
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('should match snapshot when multiple operatives', () => {
    const workOrder = {
      closedDated: '2024-09-26T12:34:56',
      paymentType: 'Bonus',
      operatives: [
        { name: 'Operative One', jobPercentage: 50 },
        { name: 'Operative Two', jobPercentage: 50 },
      ],
    }
    const note = { note: 'User comment' }

    const { asFragment } = render(
      <CompletedNoteContent
        note={note}
        workOrder={workOrder}
      />
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('should match snapshot when payment type is Bonus', () => {
    const workOrder = {
      closedDated: '2024-09-26T12:34:56',
      paymentType: 'Bonus',
      operatives: [{ name: 'Operative Name', jobPercentage: 100 }],
    }
    const note = { note: 'User comment' }

    const { asFragment } = render(
      <CompletedNoteContent
        note={note}
        workOrder={workOrder}
      />
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('should match snapshot when payment type is Close to Base', () => {
    const workOrder = {
      closedDated: '2024-09-26T12:34:56',
      paymentType: 'CloseToBase',
      operatives: [{ name: 'Operative Name', jobPercentage: 100 }],
    }
    const note = { note: 'User comment' }

    const { asFragment } = render(
      <CompletedNoteContent
        note={note}
        workOrder={workOrder}
      />
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('should match snapshot when payment type is Overtime', () => {
    const workOrder = {
      closedDated: '2024-09-26T12:34:56',
      paymentType: 'Overtime',
      operatives: [{ name: 'Operative Name', jobPercentage: 100 }],
    }
    const note = { note: 'User comment' }

    const { asFragment } = render(
      <CompletedNoteContent
        note={note}
        workOrder={workOrder}
      />
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('should match snapshot when payment type is null', () => {
    const workOrder = {
      closedDated: '2024-09-26T12:34:56',
      paymentType: null,
      operatives: [{ name: 'Operative Name', jobPercentage: 100 }],
    }
    const note = { note: 'User comment' }

    const { asFragment } = render(
      <CompletedNoteContent
        note={note}
        workOrder={workOrder}
      />
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('should match snapshot when user has uploaded photos', () => {
    const workOrder = {
      closedDated: '2024-09-26T12:34:56',
      paymentType: 'Bonus',
      operatives: [{ name: 'Operative Name', jobPercentage: 100 }],
      uploadedFileCount: {
        totalFileCount: 8,
      },
    }
    const note = { note: 'User comment' }

    const { asFragment } = render(
      <CompletedNoteContent
        note={note}
        workOrder={workOrder}
      />
    )

    expect(asFragment()).toMatchSnapshot()
  })

  describe('When further work is required', () => {
    it('should show the type of work', () => {
      const workOrder = {
        closedDated: '2024-09-26T12:34:56',
        paymentType: null,
        operatives: [{ name: 'Operative Name', jobPercentage: 100 }],
        followOnRequest: {
          id: 31,
          isSameTrade: true,
          isDifferentTrades: true,
          requiredFollowOnTrades: ['Carpentry', 'Drainage'],
          isMultipleOperatives: true,
          followOnTypeDescription: 'work required',
          stockItemsRequired: false,
          nonStockItemsRequired: false,
          materialNotes: null,
          additionalNotes: null,
        },
      }
      const note = { note: 'User comment' }

      const { asFragment } = render(
        <CompletedNoteContent
          note={note}
          workOrder={workOrder}
        />
      )

      expect(asFragment()).toMatchSnapshot()
    })

    it('should show materials', () => {
      const workOrder = {
        closedDated: '2024-09-26T12:34:56',
        paymentType: null,
        operatives: [{ name: 'Operative Name', jobPercentage: 100 }],
        followOnRequest: {
          id: 31,
          isSameTrade: true,
          isDifferentTrades: false,
          requiredFollowOnTrades: [],
          isMultipleOperatives: false,
          followOnTypeDescription: 'work required',
          stockItemsRequired: true,
          nonStockItemsRequired: true,
          materialNotes: 'some description about materials',
          additionalNotes: null,
        },
      }
      const note = { note: 'User comment' }

      const { asFragment } = render(
        <CompletedNoteContent
          note={note}
          workOrder={workOrder}
        />
      )

      expect(asFragment()).toMatchSnapshot()
    })

    it('should show additional notes', () => {
      const workOrder = {
        closedDated: '2024-09-26T12:34:56',
        paymentType: null,
        operatives: [{ name: 'Operative Name', jobPercentage: 100 }],
        followOnRequest: {
          id: 31,
          isSameTrade: true,
          isDifferentTrades: false,
          requiredFollowOnTrades: [],
          isMultipleOperatives: false,
          followOnTypeDescription: 'work required',
          stockItemsRequired: false,
          nonStockItemsRequired: false,
          materialNotes: null,
          additionalNotes: 'Some additonal notes about the work required',
        },
      }
      const note = { note: 'User comment' }

      const { asFragment } = render(
        <CompletedNoteContent
          note={note}
          workOrder={workOrder}
        />
      )

      expect(asFragment()).toMatchSnapshot()
    })
  })
})
