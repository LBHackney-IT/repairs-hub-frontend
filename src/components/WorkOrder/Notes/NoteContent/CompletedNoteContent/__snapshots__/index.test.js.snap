// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompletedNoteContent component When further work is required should show additional notes 1`] = `
<DocumentFragment>
  <span>
    Closed - Completed: 26 September 2024 at 12:34 - User comment - Assigned operatives Operative Name: 100%
    <br />
  </span>
  <div
    style="background: rgb(238, 238, 238); padding: 15px;"
  >
    <span>
      <strong
        style="font-weight: 700; color: rgb(51, 51, 51);"
      >
        Details of further work required
      </strong>
       
      <br />
      <br />
    </span>
    <span>
      <strong>
        Type of work required
      </strong>
    </span>
    <ul
      class="lbh-list lbh-list--bullet lbh-body-s"
      style="color: rgb(51, 51, 51);"
    >
      <li
        style="margin-top: 5px;"
      >
        Same trade
      </li>
    </ul>
    <br />
    <span
      style="color: rgb(51, 51, 51);"
    >
      work required
      <br />
    </span>
    <br />
    <span>
      <strong>
        Additional notes
      </strong>
      <br />
    </span>
    <span
      style="color: rgb(51, 51, 51);"
    >
      Some additonal notes about the work required
    </span>
    <br />
  </div>
</DocumentFragment>
`;

exports[`CompletedNoteContent component When further work is required should show materials 1`] = `
<DocumentFragment>
  <span>
    Closed - Completed: 26 September 2024 at 12:34 - User comment - Assigned operatives Operative Name: 100%
    <br />
  </span>
  <div
    style="background: rgb(238, 238, 238); padding: 15px;"
  >
    <span>
      <strong
        style="font-weight: 700; color: rgb(51, 51, 51);"
      >
        Details of further work required
      </strong>
       
      <br />
      <br />
    </span>
    <span>
      <strong>
        Type of work required
      </strong>
    </span>
    <ul
      class="lbh-list lbh-list--bullet lbh-body-s"
      style="color: rgb(51, 51, 51);"
    >
      <li
        style="margin-top: 5px;"
      >
        Same trade
      </li>
    </ul>
    <br />
    <span
      style="color: rgb(51, 51, 51);"
    >
      work required
      <br />
    </span>
    <br />
    <span>
      <strong>
        Materials required
      </strong>
    </span>
    <ul
      class="lbh-list lbh-list--bullet lbh-body-s"
      style="color: rgb(51, 51, 51);"
    >
      <li
        style="margin-top: 5px;"
      >
        Stock items required
      </li>
      <li
        style="margin-top: 5px;"
      >
        Non stock items required
      </li>
    </ul>
    <br />
    <span
      style="color: rgb(51, 51, 51);"
    >
      some description about materials
      <br />
    </span>
  </div>
</DocumentFragment>
`;

exports[`CompletedNoteContent component When further work is required should show the type of work 1`] = `
<DocumentFragment>
  <span>
    Closed - Completed: 26 September 2024 at 12:34 - User comment - Assigned operatives Operative Name: 100%
    <br />
  </span>
  <div
    style="background: rgb(238, 238, 238); padding: 15px;"
  >
    <span>
      <strong
        style="font-weight: 700; color: rgb(51, 51, 51);"
      >
        Details of further work required
      </strong>
       
      <br />
      <br />
    </span>
    <span>
      <strong>
        Type of work required
      </strong>
    </span>
    <ul
      class="lbh-list lbh-list--bullet lbh-body-s"
      style="color: rgb(51, 51, 51);"
    >
      <li
        style="margin-top: 5px;"
      >
        Same trade
      </li>
      <li
        style="margin-top: 5px;"
      >
        Different trade(s): Carpentry, Drainage
      </li>
      <li
        style="margin-top: 5px;"
      >
        Multiple operatives
      </li>
    </ul>
    <br />
    <span
      style="color: rgb(51, 51, 51);"
    >
      work required
      <br />
    </span>
  </div>
</DocumentFragment>
`;

exports[`CompletedNoteContent component should match snapshot when multiple operatives 1`] = `
<DocumentFragment>
  <span>
    Closed - Completed: 26 September 2024 at 12:34 - User comment - Assigned operatives Operative One: 50%, Operative Two: 50% - Bonus calculation
    <br />
  </span>
</DocumentFragment>
`;

exports[`CompletedNoteContent component should match snapshot when no operative 1`] = `
<DocumentFragment>
  <span>
    Closed - Completed: 26 September 2024 at 12:34 - User comment - Bonus calculation
    <br />
  </span>
</DocumentFragment>
`;

exports[`CompletedNoteContent component should match snapshot when one operative 1`] = `
<DocumentFragment>
  <span>
    Closed - Completed: 26 September 2024 at 12:34 - User comment - Assigned operatives Operative Name: 50% - Bonus calculation
    <br />
  </span>
</DocumentFragment>
`;

exports[`CompletedNoteContent component should match snapshot when payment type is Bonus 1`] = `
<DocumentFragment>
  <span>
    Closed - Completed: 26 September 2024 at 12:34 - User comment - Assigned operatives Operative Name: 100% - Bonus calculation
    <br />
  </span>
</DocumentFragment>
`;

exports[`CompletedNoteContent component should match snapshot when payment type is Close to Base 1`] = `
<DocumentFragment>
  <span>
    Closed - Completed: 26 September 2024 at 12:34 - User comment - Assigned operatives Operative Name: 100% - Close to base (Operative payment made)
    <br />
  </span>
</DocumentFragment>
`;

exports[`CompletedNoteContent component should match snapshot when payment type is Overtime 1`] = `
<DocumentFragment>
  <span>
    Closed - Completed: 26 September 2024 at 12:34 - User comment - Assigned operatives Operative Name - Overtime work order (SMVs not included in Bonus)
    <br />
  </span>
</DocumentFragment>
`;

exports[`CompletedNoteContent component should match snapshot when payment type is null 1`] = `
<DocumentFragment>
  <span>
    Closed - Completed: 26 September 2024 at 12:34 - User comment - Assigned operatives Operative Name: 100%
    <br />
  </span>
</DocumentFragment>
`;

exports[`CompletedNoteContent component should match snapshot when user comment is empty 1`] = `
<DocumentFragment>
  <span>
    Closed - Completed: 26 September 2024 at 12:34 - Assigned operatives Operative Name: 100% - Bonus calculation
    <br />
  </span>
</DocumentFragment>
`;

exports[`CompletedNoteContent component should match snapshot when user comment is not empty 1`] = `
<DocumentFragment>
  <span>
    Closed - Completed: 26 September 2024 at 12:34 - User comment - Assigned operatives Operative Name: 100% - Bonus calculation
    <br />
  </span>
</DocumentFragment>
`;
